File: app/store.py
"""
Demo in-memory user store.
NOTE (flagged in improvement plan): replace with a real database
(e.g. Postgres via SQLAlchemy) before production use. Data does not persist
across restarts and this is not safe for concurrent multi-process deployment.
"""
from dataclasses import dataclass


@dataclass
class User:
    username: str
    hashed_password: str


class UserStore:
    def __init__(self) -> None:
        self._users: dict[str, User] = {}

    def get(self, username: str) -> User | None:
        return self._users.get(username)

    def exists(self, username: str) -> bool:
        return username in self._users

    def create(self, username: str, hashed_password: str) -> User:
        user = User(username=username, hashed_password=hashed_password)
        self._users[username] = user
        return user


user_store = UserStore()

File: src/main/java/app/Store.java
package app;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Demo in-memory user store.
 * NOTE (flagged in improvement plan): replace with a real database
 * (e.g. Postgres) before production use. Data does not persist across
 * restarts.
 */
public final class Store {

    public record User(String username, String hashedPassword) {
    }

    private final Map<String, User> users = new ConcurrentHashMap<>();

    public User get(String username) {
        if (username == null) {
            return null;
        }
        return users.get(username);
    }

    public boolean exists(String username) {
        return username != null && users.containsKey(username);
    }

    public User create(String username, String hashedPassword) {
        User user = new User(username, hashedPassword);
        users.put(username, user);
        return user;
    }

    public void clear() {
        users.clear();
    }
}
